ecommerce/core/utils.py

```diff
@@ -72,3 +79,300 @@ def use_read_replica_if_available(queryset):
+    try:
+        response = get_catalog_course_runs(site=partner.default_site, query=query, limit=limit, offset=0)
+        results = response['results']
+        course_ids = [result['key'] for result in results]
+        return course_ids
+    except (ReqConnectionError, RequestException, Timeout):
+        logger.error('Unable to connect to Catalog API.')
+        return []
+
+
+def create_predicate_from_course_run(course_rns, all_changed_operators):
+    # Remove duplicates and sort for consistency
+    unique_courses = sorted(set(course_rns))
+
+    # Properly format the output string
+    formatted_courses = ', '.join(f'"{course}"' for course in unique_courses)
+    if len(all_changed_operators) > 0:
+        return f"variant.key not in ({formatted_courses})"
+    # Return the desired output
+    return f"variant.key in ({formatted_courses})"
+
+
+def concat_operator(operator):
+    return f' {operator} ' if (operator is not None) else ''
+
+
+def process_query_string(query, site_configuration):
+    predicate = ''
+    components = split_components_on_operator(query)
+
+    # Mapping of query types to processing functions
+    query_handlers_with_site_config = {
+        'number': process_number,
+        'key': process_key,
+    }
+    query_handlers_without_site_config = {
+        'org': process_org_values,
+        'start': process_date,
+    }
+
+    for comp in components:
+        component = comp['component']
+        operator = comp['operator']
+
+        query_type = detect_string_type(component)
+
+        if query_type in query_handlers_with_site_config:
+            predicate += query_handlers_with_site_config[query_type](component, site_configuration)
+            predicate += concat_operator(operator)
+
+        elif query_type in query_handlers_without_site_config:
+            predicate += query_handlers_without_site_config[query_type](component)
+            predicate += concat_operator(operator)
+
+        else:
+            predicate = 'No match found'
+
+    return predicate
+
+
+def process_date(component):
+    component = component.replace('"', '')
+
+    return convert_date_range_to_predicate(component)
+
+
+def convert_date_range_to_predicate(date_range):
+    # Regular expression to extract the start and end dates
+    match = re.search(r'start:\[(\d{4}-\d{2}-\d{2}) TO (\d{4}-\d{2}-\d{2})\]', date_range)
+
+    if not match:
+        return ""
+
+    start_date, end_date = match.groups()
+
+    # Construct the desired predicate
+    predicate = (
+        f'attributes.`courserun-start` >= "{start_date}" '
+        f'AND attributes.`courserun-start` <= "{end_date}"'
+    )
+
+    return predicate
+
+
+def process_number(component, site_configuration):
+    if component in KEY_TO_PREDICATE_DIC:
+        return KEY_TO_PREDICATE_DIC[component]
+
+    course_ids = fetch_catalog_course_runs(component, 10, site_configuration)
+    if course_ids:
+        return process_number_value(course_ids)
+    return ''
```

Comment:
If we only require first object then we can also fetch only 1 item
```suggestion
    course_ids = fetch_catalog_course_runs(component, 1, site_configuration)
    if course_ids:
        return process_number_value(course_ids)
    return ''
```

Reply:
Done